### PR TITLE
KIP 368: SASL Reauthentication

### PR DESCRIPTION
**Pull request by vctoriawu**
Added SASL reauthentication for the following feature: [KIP-368](https://cwiki.apache.org/confluence/display/KAFKA/KIP-368%3A+Allow+SASL+Connections+to+Periodically+Re-Authenticate)

### CODE REVIEW

**Comment by showuon** [comment]
@edenhill , could you please take a look? Thank you.

**Comment by showuon** [comment]
@julesbovet , please also take a look. Thank you.

File: src/rdkafka_broker.c [hunk 1]
```diff
@@ -2274,6 +2373,13 @@ static void rd_kafka_broker_connect_auth(rd_kafka_broker_t *rkb) {
                                 : RD_KAFKA_BROKER_STATE_AUTH_LEGACY);
                         rd_kafka_broker_unlock(rkb);
 
+                        // Check if reauth is in progress, if so then client close
+                        if (rkb->rkb_rk->rk_conf.sasl.enable_refresh) {
+                                if (rkb->rkb_sasl.session_lifetime_ms > 0) {
```
Comment (showuon) [inline src/rdkafka_broker.c:2378]: nit: `if (rkb->rkb_rk->rk_conf.sasl.enable_refresh && rkb->rkb_sasl.session_lifetime_ms > 0)`

File: src/rdkafka_conf.c [hunk 2]
```diff
@@ -926,6 +926,12 @@ static const struct rd_kafka_property rd_kafka_properties[] = {
      "This builtin handler should only be used for development "
      "or testing, and not in production.",
      0, 1, 0, _UNSUPPORTED_OAUTHBEARER},
+    {_RK_GLOBAL, "enable.sasl.token.refresh", _RK_C_BOOL,
+     _RK(sasl.enable_refresh),
+     "Enable token refresh for SASL connections. "
+     "This verification can be extended by the application by "
+     "implementing a oauthbearer_token_refresh_cb.",
```
Comment (showuon) [inline src/rdkafka_conf.c:933]: nit: by implementing oauthbearer_token_refresh_cb callback. (remove `a`, add `callback`)

**Review by showuon — COMMENTED** [review]
@vctoriawu , thanks for the PR. This is a very important feature to get supported! Sorry, I'm not quite familiar with the code base. Left some minor comments. Thanks.

**Comment by vctoriawu** [comment]
@edenhill any updates on the additional requirements?

File: src/rdkafka_conf.c [hunk 3]
```diff
@@ -926,6 +926,12 @@ static const struct rd_kafka_property rd_kafka_properties[] = {
      "This builtin handler should only be used for development "
      "or testing, and not in production.",
      0, 1, 0, _UNSUPPORTED_OAUTHBEARER},
+    {_RK_GLOBAL, "enable.sasl.token.refresh", _RK_C_BOOL,
```
[the remark this answers is not in the document]
Comment (SamBarker) [reply src/rdkafka_conf.c:929]: I've no insight into the need for the flag but I do think re-authentication is generally valid.

The intention of re-auth is to ensure that the client permissions are still valid so it is useful across all SASL auth types.

Every re-auth period the PLAIN and SCRAM tokens would be re-presented (unchanged on the client side) to allow the server to re-validate the session. If the account represented by the token had been suspended the re-auth call will fail and thus the session will be terminated.

**Comment by showuon** [comment]
@edenhill , any update? Thanks.

**Comment by showuon** [comment]
@edenhill , any update for the requirements?

**Comment by k-wall** [comment]
Any way we can help with this feature?

**Comment by edenhill** [comment]
I think the proposed solution is overly complex, and I don't think the special handling of the reauth requests is really needed.

Can't we get away with just sending the SaslHandshakeRequest upon reauth timer expiry and handling the results on response; if reauth succeeds nothing needs to be done (just restart the timer), if it fails we close the broker connection and perhaps raise a fatal error.

There's also the addition of sasl_set_credentials() in #4033 that is also needed for reauth to be meaningful.

**Comment by vctoriawu** [comment]
I agree that that could be a potential solution, but reading the description of the KIP leads me to believe that this is a requirement to hold back other requests:

"the broker will communicate the session expiration time as part of the final SASL_AUTHENTICATE response.  If this value is positive, then the client will automatically re-authenticate before anything else unrelated to re-authentication is sent beyond that expiration point."

also:

"Note also that the client cannot queue up additional send requests beyond the one that triggers re-authentication to occur until re-authentication succeeds and the triggering one is sent."

Not sure if I am misinterpreting

**Comment by milindl** [comment]
Thanks a lot @vctoriawu for starting this.

We discussed this quite a bit internally, and we decided to do it with the new broker state as Magnus suggested in the comment [above](https://github.com/confluentinc/librdkafka/pull/3754#pullrequestreview-933416062). The approach does take care of the case you mention [here.](https://github.com/confluentinc/librdkafka/pull/3754#issuecomment-1411572826). 

The PR #4301 is a draft of that approach. I've taken the commits from this PR and built upon those. I've also added a few tests. (The credential refresh callback is yet to be implemented, though, it's only the reauthentication bit -- any credentials, if the user wants to reset them, have to be done on their own for SASL PLAIN and SCRAM.).

If you have any comment or question about this approach, please send them to that PR.

**Comment by milindl** [comment]
Closing this with the 2.2.0 release which has this feature. Thanks

**Pull request closed without merging**